### PR TITLE
support qnn mean (dim=None)

### PR DESCRIPTION
**Pull request by cccclai**
Summary: Address mean op lower failure. When dim is not specified, it will take mean across all axes. For QNN, we need to get axes based on input shape

Differential Revision: D83520776

### CODE REVIEW

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 1]
```diff
@@ -1025,6 +1025,13 @@ def test_qnn_backend_mean_dim(self):
             with self.subTest(i=i):
                 self.lower_module_and_test_output(module, sample_input)
 
+    def test_qnn_backend_mean(self):
+        modules = [Mean(), Mean()]  # noqa: F405
```
Comment (haowhsu-quic) [inline backends/qualcomm/tests/test_qnn_delegate.py]: Do you need more configurations here? like `Mean(dim=0)`, `Mean(dim=0, keepdim=True)`.

File: backends/qualcomm/tests/models.py [hunk 2]
```diff
@@ -1348,6 +1348,22 @@ def forward(self, x):
         return torch.mean(x, (-1, -2))
 
 
+class Mean(torch.nn.Module):
```
Comment (haowhsu-quic) [inline backends/qualcomm/tests/models.py:1336]: Could you help remove the legacy models / tests for `MeanWOKeppDim`, `MeanWKeppDim`? thanks.

Comment (cccclai) [reply backends/qualcomm/tests/models.py:1336]: Yes updated

File: backends/qualcomm/tests/test_qnn_delegate.py (hunk 1 above)
Comment (cccclai) [reply backends/qualcomm/tests/test_qnn_delegate.py]: updated

**Comment by cccclai** [comment]
Does it look good? I think it can fix 5 failing op tests

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 3]
```diff
@@ -2666,13 +2710,59 @@ def test_qnn_backend_max_pool2d(self):
...
+
+        for i, test in enumerate(test_comb):
             with self.subTest(i=i):
-                module = self.get_qdq_module(module, sample_input)
-                self.lower_module_and_test_output(module, sample_input)
+                print("test i", i)
+                module = self.get_qdq_module(
+                    test[QCOM_MODULE], test[QCOM_SAMPLE_INPUTS]
+                )
+                module = self.get_qdq_module(module, test[QCOM_SAMPLE_INPUTS])
```
Comment (haowhsu-quic) [inline backends/qualcomm/tests/test_qnn_delegate.py]: Might need to remove L2759~2762, thanks.

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 4]
```diff
@@ -1018,12 +1018,56 @@ def test_qnn_backend_max_pool2d(self):
...
+            {
+                QCOM_MODULE: Mean(dim=(1, 3), keepdim=True),
+                QCOM_SAMPLE_INPUTS: (torch.randn(2, 3, 4, 5, 6),),
+            },
+        ]
+
+        for i, test in enumerate(test_comb):
             with self.subTest(i=i):
-                self.lower_module_and_test_output(module, sample_input)
+                module = self.get_qdq_module(
```
Comment (haowhsu-quic) [inline backends/qualcomm/tests/test_qnn_delegate.py]: I think we don't need `get_qdq_module` for floating point test.

File: backends/qualcomm/tests/test_qnn_delegate.py (hunk 3 above)
Comment (cccclai) [reply backends/qualcomm/tests/test_qnn_delegate.py]: Good catch, thank you for the review

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 5]
```diff
@@ -1018,12 +1018,53 @@ def test_qnn_backend_max_pool2d(self):
...
+            },
+            # Default: reduce all dims
+            {
+                QCOM_MODULE: Mean(),
+                QCOM_SAMPLE_INPUTS: (torch.randn(10, 10),),
+            },
+            # Scalar case
+            {
+                QCOM_MODULE: Mean(),
+                QCOM_SAMPLE_INPUTS: (torch.tensor([5.0]),),
```
Comment (chenweng-quic) [inline backends/qualcomm/tests/test_qnn_delegate.py]: Should it be torch.tensor(5.0) if you want to test scalar?

File: backends/qualcomm/builders/op_mean_dim.py [hunk 6]
```diff
@@ -40,7 +40,16 @@ def define_node(
         )
 
         # mean dims and keep dims
-        mean_dims = cast(List[int], node.args[1])
+        rank = len(input_node.meta["val"].shape)
+        dim_arg = node.args[1]
+
+        if dim_arg is None:
```
Comment (chenweng-quic) [inline backends/qualcomm/builders/op_mean_dim.py]: Maybe edge case, could dim_arg be empty list?

File: backends/qualcomm/tests/test_qnn_delegate.py [hunk 7]
```diff
@@ -2666,13 +2707,55 @@ def test_qnn_backend_max_pool2d(self):
...
+                QCOM_MODULE: Mean(dim=(1, 3), keepdim=True),
+                QCOM_SAMPLE_INPUTS: (torch.randn(2, 3, 4, 5, 6),),
+            },
+        ]
+
+        for i, test in enumerate(test_comb):
             with self.subTest(i=i):
-                module = self.get_qdq_module(module, sample_input)
-                self.lower_module_and_test_output(module, sample_input)
+                module = self.get_qdq_module(module, test[QCOM_SAMPLE_INPUTS])
```
Comment (haowhsu-quic) [inline backends/qualcomm/tests/test_qnn_delegate.py]: I guess this should be `module = self.get_qdq_module(test[QCOM_MODULE], test[QCOM_SAMPLE_INPUTS])`

File: backends/qualcomm/builders/op_mean_dim.py (hunk 6 above)
Comment (cccclai) [reply backends/qualcomm/builders/op_mean_dim.py]: Yes I verify the eager mode and it does support empty list. Updated

File: backends/qualcomm/tests/test_qnn_delegate.py (hunk 5 above)
Comment (cccclai) [reply backends/qualcomm/tests/test_qnn_delegate.py]: Thanks, good catch

File: backends/qualcomm/tests/test_qnn_delegate.py (hunk 7 above)
Comment (cccclai) [reply backends/qualcomm/tests/test_qnn_delegate.py]: Sorry I should be more careful..

File: backends/qualcomm/tests/test_qnn_delegate.py (hunk 5 above)
Comment (cccclai) [reply backends/qualcomm/tests/test_qnn_delegate.py]: Actually this is indeed a good test. ReduceMean doesn't support 0d tensor, I think we need to have a pass to convert 0d to 1d tensor if the input of mean is 0d. I have some sketch code here. What do you think? For now, I comment out the test case and think we can follow up on this
```
import torch
from executorch.exir.pass_base import ExportPass, PassResult

class Rank0ToRank1(ExportPass):
    """
    For selected ops and selected input positions, if the input is rank-0 (scalar),
    insert a reshape to [1] before the op.
    """

    def __init__(self, op_input_map=None) -> None:
        super().__init__()
        # key is the op, value is the input indices to be reshaped
        self.op_input_map = {
          torch.ops.aten.mean.dim: [0],
        }

    def call(self, graph_module: torch.fx.GraphModule) -> PassResult:
        graph = graph_module.graph
        changed = False

        for node in list(graph.nodes):
            if node.op == "call_function" and node.target in self.op_input_map:
                input_indices = self.op_input_map[node.target]

                new_args = list(node.args)
                for idx in input_indices:
                    if idx < len(new_args):
                        inp = new_args[idx]
                        if hasattr(inp, "meta"):
                            val = inp.meta.get("val", None)
                            if val is not None and hasattr(val, "shape") and val.shape == ():
                                # Insert reshape right before the op
                                with graph.inserting_before(node):
                                    reshape_node = graph.call_function(
                                        torch.ops.aten.reshape.default,
                                        args=(inp, (1,))
                                    )
                                    reshape_node.meta["val"] = val.reshape(1,)

                                # Replace arg idx with reshape_node
                                new_args[idx] = reshape_node
                                changed = True

                # update node args if modified
                node.args = tuple(new_args)

        if changed:
            graph_module.recompile()
        return PassResult(graph_module, changed)

```

**Review by chenweng-quic — APPROVED** [review]

**Review by kirklandsign — APPROVED** [review]
